### PR TITLE
Fabfile cadf

### PR DESCRIPTION
**Pull request by jxstanford**
This wraps up a couple of tickets:

1) docker-based developer environment
2) CADF event generation from openstack nodes (in support of OpenTrail)
3) also fixes and issue with get_client authentication

**Docker-based Development Environment (*assumes MacOSX*)**
* pull this branch (fabfile_cadf)
* if not already using docker toolbox / docker machine, remove all vestiges of docker and virtualbox
  * brew uninstall docker, docker-compose, boot2docker, fig, etc.
  * remove virtual machines
  * remove host only networks
  * remove stray virtualbox DHCP servers for 172.24.4.1 
    * VBoxManage list dhcpservers
    * VBoxManage dhcpserver remove -ifname vboxnet{x}
  * uninstall VirtualBox
* follow process in http://goldstone-server.readthedocs.org/en/fabfile_cadf/HACKING/

**CADF event generation**

the configure_stack task in fabric sets up the RDO instance to ship events back to goldstone.  The task is run from within the init_dev_env.sh script, which you will have run when you set up the docker-based dev environment.  

You can verify that events are flowing by logging in to goldstone (http://127.0.0.1:8000) and going to the event browser screen.

### CODE REVIEW

**Comment by jxstanford** [comment]
going to do a bit more automation.  push coming later today.

**Comment by stugots** [comment]
This pull has increased pylint errors in fabfile and installer_fabfile.  Also there's no improvement in goldstone's pylint score.

File: goldstone/initial_load.py [hunk 1]
```diff
@@ -88,7 +88,22 @@ def _put_all_templates(server=settings.ES_SERVER):
...
+
+    from elasticsearch.exceptions import RequestError
+    from .models import es_conn
+
+    try:
+        conn = es_conn(server)
+        conn.indices.delete('goldstone*')
+    except RequestError:
+        print "?ERROR: Index deletion failed. Please report this!"
+        raise
```
Comment (stugots) [inline goldstone/initial_load.py]: Replace 101 and 102 with raise("? Index deletion failed. Please report this!")

File: installer_fabfile.py [hunk 2]
```diff
@@ -62,6 +63,7 @@
 # The Goldstone settings path, relative to the Goldstone root where we're
 # executing from.
 PROD_SETTINGS = "goldstone.settings.production"
+PROD_DOCKER_SETTINGS = "goldstone.settings.production"
```
Comment (stugots) [inline installer_fabfile.py]: This is unused.

**Comment by stugots** [comment]
If I were to wipe my system and do this from scratch, would I still need to edit /etc/hosts?

**Comment by jxstanford** [comment]
pylint on this branch: 9.72/10
pylint on master branch: 9.68/10

fabfiles are all 10/10

**Comment by stugots** [comment]
This is the only reference to goldstone-docker in HACKING. It should either be expanded to explain what goldstone-docker is for, or it should be deleted.

screen shot 2015-08-28 at 12 01 48 pm

**Comment by jxstanford** [comment]
all comments above have been addressed.

**Comment by rsmitty** [comment]
Seems to install correctly for me!

screen shot 2015-08-28 at 2 01 31 pm

 Here's the bit of feedback I had. I started from bare environment (removed docker, machine, compose, virtualbox, did not have fig installed)

--------------------------------------

**"Download OpenStack VM" section**

- Download ova from https://region-a.geo-1.objects.hpcloudsvc.com/v1/10815605991908/RDO-Images/RDO-kilo-201508.ova. The old HP link requires registration.

--------------------------------------

**Before "Configure VirtualBox Networking"**

- No instructions (that I saw) to create 'default' docker vm before running "configure_vbox.sh". This is what I did before running it:
 - `brew install docker docker-machine docker-compose`
 - `docker-machine create --driver virtualbox default`
 - `docker-machine upgrade default`
 - `docker-machine stop default`
- Then, I needed to remove all host networks (again, even after doing it when uninstalling vbox)
-- Virtualbox -> Preferences -> Networks -> Host-only Networks -> Delete each network
- Finally, I could run configure_vbox.sh

--------------------------------------

**Other Notes**

- It wasn't really clear to me (at first) that I needed to run two terminal sessions, one for `./bin/start_dev_env.sh` and one for `fab runserver`. They are sequential in the code block, so maybe some text indicating they should be separate. 

- Some text on sourcing environment variables and booting a vm to show the graphs working and stuff would be helpful at the end. @jxstanford gave me `nova boot --image cirros --flavor m1.tiny ceilo0` which seemed to work just fine.

- After conversation with John, we've discovered a couple of possible follow on bugs which may not be related to this deployment. 

-----------------

**Bugs**

 - I can't seem to see any API graphs in the UI. Doesn't seem to matter which screen I'm on. There are API events in the report however, just the graphs seem broken. Here's a photo to illustrate:

screen shot 2015-08-28 at 3 08 53 pm

 - There is some error about type-casting in the logs. Here's the output:
 ```
gssearch_1 | [2015-08-28 22:13:07,742][DEBUG][action.search.type       ] [f235222f7c34] All shards failed for phase: [query]
gssearch_1 | java.lang.ClassCastException: org.elasticsearch.index.fielddata.plain.PagedBytesIndexFieldData cannot be cast to org.elasticsearch.index.fielddata.IndexNumericFieldData
gssearch_1 |    at org.elasticsearch.search.aggregations.support.AggregationContext.numericField(AggregationContext.java:160)
gssearch_1 |    at org.elasticsearch.search.aggregations.support.AggregationContext.valuesSource(AggregationContext.java:137)
gssearch_1 |    at org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory.create(ValuesSourceAggregatorFactory.java:53)
gssearch_1 |    at org.elasticsearch.search.aggregations.AggregatorFactories.createAndRegisterContextAware(AggregatorFactories.java:53)
gssearch_1 |    at org.elasticsearch.search.aggregations.AggregatorFactories.createSubAggregators(AggregatorFactories.java:71)
gssearch_1 |    at org.elasticsearch.search.aggregations.Aggregator.<init>(Aggregator.java:191)
gssearch_1 |    at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.<init>(BucketsAggregator.java:39)
gssearch_1 |    at org.elasticsearch.search.aggregations.bucket.histogram.HistogramAggregator.<init>(HistogramAggregator.java:64)
gssearch_1 |    at org.elasticsearch.search.aggregations.bucket.histogram.HistogramAggregator$Factory.create(HistogramAggregator.java:182)
gssearch_1 |    at org.elasticsearch.search.aggregations.bucket.histogram.HistogramAggregator$Factory.create(HistogramAggregator.java:138)
gssearch_1 |    at org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory.create(ValuesSourceAggregatorFactory.java:54)
gssearch_1 |    at org.elasticsearch.search.aggregations.AggregatorFactories.createAndRegisterContextAware(AggregatorFactories.java:53)
gssearch_1 |    at org.elasticsearch.search.aggregations.AggregatorFactories.createTopLevelAggregators(AggregatorFactories.java:157)
gssearch_1 |    at org.elasticsearch.search.aggregations.AggregationPhase.preProcess(AggregationPhase.java:79)
gssearch_1 |    at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:100)
gssearch_1 |    at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:301)
gssearch_1 |    at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:312)
gssearch_1 |    at org.elasticsearch.search.action.SearchServiceTransportAction$5.call(SearchServiceTransportAction.java:231)
gssearch_1 |    at org.elasticsearch.search.action.SearchServiceTransportAction$5.call(SearchServiceTransportAction.java:228)
gssearch_1 |    at org.elasticsearch.search.action.SearchServiceTransportAction$23.run(SearchServiceTransportAction.java:559)
gssearch_1 |    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
gssearch_1 |    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
gssearch_1 |    at java.lang.Thread.run(Thread.java:745)
 ```

**Comment by stugots** [comment]
A full install from scratch worked on my laptop.

HACKING needs to enumerate the default accounts and passwords that the instillation creates.  A table would be fine.  I had to look in init_dev_env.sh , and the postactivate script, to discover them.

p.s. @rsmitty: Great comments!

**Comment by stugots** [comment]
Whoops, I spoke too soon.  I'm getting Elasticsearch errors on some of the Metrics graphs.

screen shot 2015-08-28 at 4 18 20 pm

**Comment by stugots** [comment]
@jxstanford:

My only remaining comment is:

HACKING should enumerate the accounts and passwords created by the installation. A table would be fine.  A customer shouldn't have to search in init_dev_env.sh and the postactivate script to discover them.

**Comment by jxstanford** [comment]
@stugots it is in the output from init_dev_env.sh.  So, if you follow the instructions, you will see those user/pw details.  Think that's sufficient?

**Comment by stugots** [comment]
@jxstanford: Ah, I missed that!  Yes that'll be sufficient. :+1:

**Comment by jxstanford** [comment]
ok, will wait for experience data from @lexjacobs

**Comment by lexjacobs** [comment]
### virtual box configuration
The port forwarding rules trigger an error in virtual box, but they seem to work anyway. But you have to click 'ok > cancel' to get out of the error popup state.

screen shot 2015-08-31 at 3 29 22 pm


Changing video RAM to 9MB made this error go away. Perhaps there is a pref that can be added to configure_vbox.sh ?

screen shot 2015-08-31 at 3 28 50 pm


### hacking.md file

line 102 no longer seems relevant: **_Note that the goldstone-server virtualenv is only meant to be run in a single terminal window._**
as the current setup requires 2 terminal windows .

### after fab runserver

api perf charts erroring as follows:
screen shot 2015-08-31 at 4 13 01 pm

http://localhost:8000/#reports/apibrowser
chart is blank. my guess is that if the api perf charts were resolving, this chart would be populated.


screen shot 2015-08-31 at 4 12 51 pm

### nits
I had to search though this PR to figure out what the new username/password was for logging in. The user/pass as listed in `description` in the RDO-kilo instance lists an out of date user/pass combo.

**Comment by lexjacobs** [comment]
Per @slashk , COMMON_RULE_LIST in configure_vbox.sh may need editing:

image

**Comment by lexjacobs** [comment]
As discussed with JS, :+1: as dev environment is up and running, and remaining issues will be ticketed.

**Comment by jxstanford** [comment]
@lexjacobs all of your instance problems (video memory, user/pass docs) have been fixed.  You can download the latest version of the RDO instance (linked in hacking) or just march on with what you have. 

I was not able to recreate the API chart issues in my latest runs, but we should keep an eye on it.  I have seen it before.  

Removed the additional NAT rules that were unnecessary, per @slashk suggestion.

Did a bit more doc tidying, and pulling the trigger...

**Comment by lexjacobs** [comment]
@jxstanford 
downloaded latest version of RDO instance, as mentioned above.
From link: https://region-a.geo-1.objects.hpcloudsvc.com/v1/10815605991908/RDO-Images/RDO-kilo-201508.ova

Just a heads up the video memory issue was in the "default" virtual machine, not the RDO-kilo instance.
And the description of the RDO-kilo virtual machine is still showing the outdated user/password after downloading the latest file that is linked in hacking.md.

**Comment by lexjacobs** [comment]
@jxstanford confirming new ./bin/configure_vbox.sh worked and removed the error documented above.

**Pull request merged**